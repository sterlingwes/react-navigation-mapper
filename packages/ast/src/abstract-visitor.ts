import ts from "typescript";

export interface AbstractVisitorOptions<GlobalState, ModuleState> {
  initialGlobalState?: GlobalState;
  initialModuleState?: ModuleState;
}

// TODO: add getAndReset moduleState which can be called after visiting a whole file
// TODO: add getAndReset globalState whcih can be called after visiting a whole project

export abstract class AbstractVisitor<
  GlobalState extends Record<any, any>,
  ModuleState extends Record<any, any>,
> {
  private options: {
    globalState: GlobalState;
    moduleState: ModuleState;
  };

  constructor(options: AbstractVisitorOptions<GlobalState, ModuleState>) {
    this.options = {
      globalState: options.initialGlobalState ?? ({} as Record<any, any>),
      moduleState: options.initialModuleState ?? ({} as Record<any, any>),
    };
  }

  get globalState() {
    return this.options.globalState;
  }

  get moduleState() {
    return this.options.moduleState;
  }

  get context(): {
    globalState: GlobalState;
    moduleState: ModuleState;
  } {
    return {
      globalState: this.globalState,
      moduleState: this.moduleState,
    };
  }

  visit(node: ts.Node) {
    this.switchNode(node);
  }

  abstract cases: Array<
    readonly [
      (node: any) => any,
      (
        | ((
            node: any,
            context: { globalState: GlobalState; moduleState: ModuleState }
          ) => any)
        | {
            onEnter?: (
              node: any,
              context: { globalState: GlobalState; moduleState: ModuleState }
            ) => any;
            onExit?: (
              node: any,
              context: { globalState: GlobalState; moduleState: ModuleState }
            ) => any;
          }
      ),
    ]
  >;

  case<T extends ts.Node>(
    filter: (node: ts.Node) => node is T,
    method:
      | ((
          node: T,
          context: { globalState: GlobalState; moduleState: ModuleState }
        ) => any)
      | {
          onEnter?: (
            node: T,
            context: { globalState: GlobalState; moduleState: ModuleState }
          ) => any;
          onExit?: (
            node: T,
            context: { globalState: GlobalState; moduleState: ModuleState }
          ) => any;
        }
  ) {
    return [filter, method] as const;
  }

  private switchNode(node: ts.Node) {
    const matches = this.cases.filter(([filter]) => {
      return filter(node);
    });

    const exitMethods: Array<
      (
        node: any,
        context: { globalState: GlobalState; moduleState: ModuleState }
      ) => any
    > = [];

    matches.forEach((match) => {
      if (match && match[1]) {
        const [, method] = match;
        if (typeof method === "function") {
          method(node, this.context);
        } else {
          const { onEnter, onExit } = method;
          if (onEnter) {
            onEnter(node, this.context);
          }
          if (onExit) {
            exitMethods.push(onExit);
          }
        }
      }
    });

    ts.forEachChild(node, (childNode) => this.switchNode(childNode));
    exitMethods.forEach((method) => method(node, this.context));
  }
}
