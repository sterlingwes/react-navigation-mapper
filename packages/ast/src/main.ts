import type { SourceFile } from "typescript";
import { NavigationDetectorVisitor } from "./visitors/navigation-detector";
import { NavigatorDetectorVisitor } from "./visitors/navigator-detector";

export const parseFile = (sourceFile: SourceFile) => {
  // TODO: allow for running through the file once with multiple visitors

  const navigationVisitor = new NavigationDetectorVisitor();
  navigationVisitor.visit(sourceFile);
  const { components } = navigationVisitor.globalState;

  const navigatorVisitor = new NavigatorDetectorVisitor();
  navigatorVisitor.visit(sourceFile);
  const { stacks } = navigatorVisitor.globalState;
  const skippable = Object.keys(stacks).length === 0 && components.length === 0;

  return { components, stacks, source: sourceFile.fileName, skippable };
};
