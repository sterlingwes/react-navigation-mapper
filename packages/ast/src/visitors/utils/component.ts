import ts from "typescript";

export const varDeclarationIsPossibleFunctionComponent = (
  node: ts.VariableDeclaration
) => {
  return (
    node.initializer?.kind === ts.SyntaxKind.ArrowFunction ||
    node.initializer?.kind === ts.SyntaxKind.FunctionExpression
  );
};

export const getVarDeclarationName = (node: ts.VariableDeclaration) => {
  return (node.name as ts.Identifier).escapedText as string;
};

export const getBlockPosition = (node: ts.VariableDeclaration) => {
  if (node.initializer?.kind === ts.SyntaxKind.ArrowFunction) {
    return (node.initializer as ts.ArrowFunction).body.pos;
  }

  throw new Error("getBlockPosition: Only arrow functions are supported");
};

export const returnsJSX = (node: ts.ReturnStatement) => {
  return (
    node.expression?.kind === ts.SyntaxKind.JsxElement ||
    (node.expression?.kind === ts.SyntaxKind.ParenthesizedExpression &&
      (node.expression as ts.ParenthesizedExpression).expression.kind ===
        ts.SyntaxKind.JsxElement)
  );
};
