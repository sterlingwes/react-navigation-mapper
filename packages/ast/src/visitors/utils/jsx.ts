import {
  isArrowFunction,
  isIdentifier,
  isJsxAttribute,
  isJsxExpression,
  isObjectLiteralExpression,
  isParenthesizedExpression,
  isStringLiteral,
  type JsxAttribute,
  type Node,
} from "typescript";

export const getJsxAttributeWithStringValue = (attr: Node) => {
  if (
    isJsxAttribute(attr) &&
    attr.initializer &&
    isStringLiteral(attr.initializer)
  ) {
    return attr.initializer.text;
  }
};

export const getJsxAttributeWithObjectLiteralValue = (attr: Node) => {
  if (
    isJsxAttribute(attr) &&
    attr.initializer &&
    isJsxExpression(attr.initializer) &&
    attr.initializer.expression &&
    isObjectLiteralExpression(attr.initializer.expression)
  ) {
    return attr.initializer.expression;
  }
};

export const getJsxAttributeWithIdentifierValueAsString = (attr: Node) => {
  if (
    isJsxAttribute(attr) &&
    attr.initializer &&
    isJsxExpression(attr.initializer) &&
    attr.initializer.expression &&
    isIdentifier(attr.initializer.expression)
  ) {
    return attr.initializer.expression.escapedText as string;
  }
};

export const getJsxAttributeWithArrowFunctionValue = (attr: Node) => {
  // TODO handle case with return statement vs. inline return
  if (
    isJsxAttribute(attr) &&
    attr.initializer &&
    isJsxExpression(attr.initializer) &&
    attr.initializer.expression &&
    isArrowFunction(attr.initializer.expression) &&
    isParenthesizedExpression(attr.initializer.expression.body) &&
    isObjectLiteralExpression(attr.initializer.expression.body.expression)
  ) {
    return attr.initializer.expression.body.expression;
  }
};
