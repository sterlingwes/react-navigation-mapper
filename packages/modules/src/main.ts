import { createProgramLookup } from "./program";

export { createProgramLookup };
