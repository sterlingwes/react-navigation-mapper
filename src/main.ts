import putout from "putout";
import * as prettier from "prettier";
import fs from "fs";

export const shakeFile = async (filePath: string) => {
  console.log(`Shaking ${filePath}`);
  const content = fs.readFileSync(filePath, "utf-8");
  const { code } = putout(content, {
    isTS: true,
    plugins: [
      ["remove-imports", require("./rules/remove-imports").default],
      [
        "unwrap-undeclared-hooks",
        require("./rules/unwrap-react-hooks").default,
      ],
      [
        "remove-undeclared-var-calls",
        require("./rules/remove-undeclared-var-calls").default,
      ],
      [
        "remove-unreachable-condition",
        require("./rules/remove-unreachable-condition").default,
      ],
      [
        "remove-undeclared-types",
        require("./rules/remove-undeclared-types").default,
      ],

      "remove-unused-variables",
      "remove-unused-expressions",
      "remove-unreachable-code",
      "remove-unreferenced-variables",
    ],
  });

  return prettier.format(code, { parser: "typescript" });
};
