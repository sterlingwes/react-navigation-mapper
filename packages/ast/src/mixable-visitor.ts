import { AbstractVisitor } from "./abstract-visitor";

export abstract class MixableVisitor<
  GlobalState extends Record<any, any>,
  ModuleState extends Record<any, any>,
> extends AbstractVisitor<GlobalState, ModuleState> {
  abstract get mixinCases(): (typeof AbstractVisitor.prototype)["cases"];
}
