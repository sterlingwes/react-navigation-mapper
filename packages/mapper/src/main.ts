import * as path from "path";
import { parseFile } from "@react-navigation-mapper/ast";
import { createProgramLookup } from "@react-navigation-mapper/modules";
import { logger } from "./utils/logger";

const rootPath = path.resolve(__dirname, "../../../");

export const mapProject = (projectPath: string) => {
  const fileMaps: Array<ReturnType<typeof parseFile>> = [];
  const { files } = createProgramLookup(projectPath, {
    rootPath,
    onSourceFile: (sourceFile) => {
      logger.debug(`onSourceFile(${sourceFile.fileName})`);
      const parsedFile = parseFile(sourceFile);
      if (!parsedFile.skippable) {
        fileMaps.push(parsedFile);
      }
    },
  });

  return { files, fileMaps };
};
