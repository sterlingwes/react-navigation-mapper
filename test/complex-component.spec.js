import { describe, it, expect } from "bun:test";

import { shakeFile } from "../src/main";
import { expectNoTypeScriptErrors } from "./utils";

describe("complex component w/ nav call", () => {
  it("should remove everything except the component function and nav call", async () => {
    const code = await shakeFile("fixtures/complex-component.tsx");
    expect(code).toMatchSnapshot();
    expectNoTypeScriptErrors(code);
  });
});
