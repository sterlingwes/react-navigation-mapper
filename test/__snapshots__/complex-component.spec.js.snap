// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`complex component w/ nav call should remove everything except the component function and nav call 1`] = `
"// @ts-nocheck
import { Image, StyleSheet, TouchableOpacity, View } from "react-native";
import { useNavigation } from "@react-navigation/native";
import { NativeStackNavigationProp } from "@react-navigation/native-stack";

export const InlineReply = ({
  inReplyToId,
}: {
  inReplyToId: string;
  replyToHandles: string[];
  onlyReply?: boolean;
  onSent?: () => void;
}) => {
  const navigation = useNavigation<NativeStackNavigationProp<any>>();

  const onPressReply = () => {
    setTimeout(() => {
      inputRef.current?.focus();
    }, 10);
  };

  const onLeaveInput = () => {};

  return (
    <View style={styles.container}>
      <Box pb={25}>
        <ReplyLine visible height={15} />
        {activeUser && (
          <Image
            source={{
              uri: activeUser?.avatar_static,
            }}
            style={[styles.userAvatar, replying && styles.userAvatarReplying]}
          />
        )}
      </Box>
      <View style={styles.rightCol}>
        {replying ? (
          <>
            <Type style={styles.yourReply} scale="S" semiBold>
              Your reply:
            </Type>
            <Input
              ref={(ref) => (inputRef.current = ref)}
              scale="S"
              style={styles.input}
              placeholder="Say something!"
              onChangeText={setTextValue}
              onBlur={onLeaveInput}
              value={textValue}
              multiline
            />
          </>
        ) : (
          <>
            <TouchableOpacity
              activeOpacity={0.5}
              style={styles.replyBtn}
              onPress={onPressReply}
            >
              <Type color={getColor("baseAccent")}>Reply</Type>
            </TouchableOpacity>
            <ExternalLink
              style={styles.externalLinkIconButton}
              color={getColor("primary")}
              onPress={() =>
                navigation.navigate("Compose", {
                  inReplyToId,
                  routeTime: Date.now(),
                })
              }
            />
          </>
        )}
      </View>
    </View>
  );
};
"
`;
